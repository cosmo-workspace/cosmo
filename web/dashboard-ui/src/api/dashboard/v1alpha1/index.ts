/* tslint:disable */
/* eslint-disable */
/**
 * Cosmo Dashboard API
 * Manipulate cosmo dashboard resource API
 *
 * The version of the OpenAPI document: v1alpha1
 * Contact: jlandowner8@gmail.com
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


export * from "./api";
export * from "./configuration";

