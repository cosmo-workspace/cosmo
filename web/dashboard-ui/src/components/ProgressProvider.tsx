import { Backdrop, CircularProgress } from '@mui/material';
import React, { createContext, useContext, useState } from 'react';

const DispatchContext = createContext<React.Dispatch<React.SetStateAction<number>>>(undefined as any);

/**
 * provider
 */
export const ProgressProvider: React.FC = ({ children }) => {
  const [count, setCount] = useState(0);
  return (<>
    <DispatchContext.Provider value={setCount}>
      {children}
    </DispatchContext.Provider>
    <Backdrop sx={{ zIndex: (theme) => theme.zIndex.drawer + 1000 }} open={count > 0}>
      <CircularProgress />
    </Backdrop>
  </>);
}

export function useProgress() {
  const setCount = useContext(DispatchContext);
  return {
    setMask: () => setCount(count => { console.log('setMask', count + 1); return count + 1 }),
    releaseMask: () => setCount(count => { console.log('releaseMask', count); return count > 0 ? count - 1 : 0 }),
  }
}