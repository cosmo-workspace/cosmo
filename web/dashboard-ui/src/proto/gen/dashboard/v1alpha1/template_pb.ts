//
//Cosmo Dashboard API
//Manipulate cosmo dashboard resource API

// @generated by protoc-gen-es v1.2.0 with parameter "target=ts"
// @generated from file dashboard/v1alpha1/template.proto (package dashboard.v1alpha1, syntax proto3)
/* eslint-disable */
// @ts-nocheck

import type { BinaryReadOptions, FieldList, JsonReadOptions, JsonValue, PartialMessage, PlainMessage } from "@bufbuild/protobuf";
import { Message, proto3 } from "@bufbuild/protobuf";

/**
 * @generated from message dashboard.v1alpha1.TemplateRequiredVars
 */
export class TemplateRequiredVars extends Message<TemplateRequiredVars> {
  /**
   * @generated from field: string var_name = 1;
   */
  varName = "";

  /**
   * @generated from field: string default_value = 2;
   */
  defaultValue = "";

  constructor(data?: PartialMessage<TemplateRequiredVars>) {
    super();
    proto3.util.initPartial(data, this);
  }

  static readonly runtime: typeof proto3 = proto3;
  static readonly typeName = "dashboard.v1alpha1.TemplateRequiredVars";
  static readonly fields: FieldList = proto3.util.newFieldList(() => [
    { no: 1, name: "var_name", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "default_value", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ]);

  static fromBinary(bytes: Uint8Array, options?: Partial<BinaryReadOptions>): TemplateRequiredVars {
    return new TemplateRequiredVars().fromBinary(bytes, options);
  }

  static fromJson(jsonValue: JsonValue, options?: Partial<JsonReadOptions>): TemplateRequiredVars {
    return new TemplateRequiredVars().fromJson(jsonValue, options);
  }

  static fromJsonString(jsonString: string, options?: Partial<JsonReadOptions>): TemplateRequiredVars {
    return new TemplateRequiredVars().fromJsonString(jsonString, options);
  }

  static equals(a: TemplateRequiredVars | PlainMessage<TemplateRequiredVars> | undefined, b: TemplateRequiredVars | PlainMessage<TemplateRequiredVars> | undefined): boolean {
    return proto3.util.equals(TemplateRequiredVars, a, b);
  }
}

/**
 * @generated from message dashboard.v1alpha1.Template
 */
export class Template extends Message<Template> {
  /**
   * @generated from field: string name = 1;
   */
  name = "";

  /**
   * @generated from field: string description = 2;
   */
  description = "";

  /**
   * @generated from field: repeated dashboard.v1alpha1.TemplateRequiredVars required_vars = 3;
   */
  requiredVars: TemplateRequiredVars[] = [];

  /**
   * @generated from field: optional bool is_default_user_addon = 4;
   */
  isDefaultUserAddon?: boolean;

  /**
   * @generated from field: bool is_cluster_scope = 5;
   */
  isClusterScope = false;

  constructor(data?: PartialMessage<Template>) {
    super();
    proto3.util.initPartial(data, this);
  }

  static readonly runtime: typeof proto3 = proto3;
  static readonly typeName = "dashboard.v1alpha1.Template";
  static readonly fields: FieldList = proto3.util.newFieldList(() => [
    { no: 1, name: "name", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "description", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "required_vars", kind: "message", T: TemplateRequiredVars, repeated: true },
    { no: 4, name: "is_default_user_addon", kind: "scalar", T: 8 /* ScalarType.BOOL */, opt: true },
    { no: 5, name: "is_cluster_scope", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
  ]);

  static fromBinary(bytes: Uint8Array, options?: Partial<BinaryReadOptions>): Template {
    return new Template().fromBinary(bytes, options);
  }

  static fromJson(jsonValue: JsonValue, options?: Partial<JsonReadOptions>): Template {
    return new Template().fromJson(jsonValue, options);
  }

  static fromJsonString(jsonString: string, options?: Partial<JsonReadOptions>): Template {
    return new Template().fromJsonString(jsonString, options);
  }

  static equals(a: Template | PlainMessage<Template> | undefined, b: Template | PlainMessage<Template> | undefined): boolean {
    return proto3.util.equals(Template, a, b);
  }
}

