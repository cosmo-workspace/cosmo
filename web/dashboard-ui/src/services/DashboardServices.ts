import { createPromiseClient } from "@bufbuild/connect";
import { createConnectTransport, createGrpcWebTransport } from "@bufbuild/connect-web";
import { useMemo } from "react";
import { AuthService } from "../proto/gen/dashboard/v1alpha1/auth_service_connectweb";
import { TemplateService } from "../proto/gen/dashboard/v1alpha1/template_service_connectweb";
import { UserService } from "../proto/gen/dashboard/v1alpha1/user_service_connectweb";
import { WorkspaceService } from "../proto/gen/dashboard/v1alpha1/workspace_service_connectweb";

const transportX = createConnectTransport({
    baseUrl: import.meta.env.BASE_URL,
});

const transport = createGrpcWebTransport({
    baseUrl: import.meta.env.BASE_URL,
});

export function useAuthService() {
    return useMemo(() => createPromiseClient(AuthService, transport), [AuthService]);
}
export function useTemplateService() {
    return useMemo(() => createPromiseClient(TemplateService, transport), [TemplateService]);
}
export function useUserService() {
    return useMemo(() => createPromiseClient(UserService, transport), [UserService]);
}

export function useWorkspaceService() {
    return useMemo(() => createPromiseClient(WorkspaceService, transport), [WorkspaceService]);
}