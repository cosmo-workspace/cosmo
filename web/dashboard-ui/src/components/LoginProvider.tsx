import { Box, CircularProgress, CssBaseline, Stack, Typography } from '@mui/material';
import { useSnackbar } from 'notistack';
import React, { createContext, useContext, useEffect, useState } from 'react';
import { User } from '../proto/gen/dashboard/v1alpha1/user_pb';
import { useAuthService, useUserService } from '../services/DashboardServices';
import { useProgress } from './ProgressProvider';


/**
 * context
 * ref: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/24509#issuecomment-774430643
 */
const Context = createContext<ReturnType<typeof useLoginModule>>(undefined as any);

/**
 * hooks
 */
export function useLogin() {
  return useContext(Context);
}

const useLoginModule = () => {
  console.log('useLoginModule');
  const [loginUser, setLoginUser] = useState<User>();
  const { enqueueSnackbar } = useSnackbar();
  const { setMask, releaseMask } = useProgress();
  const userService = useUserService();
  const authService = useAuthService();


  /**
   * verifyLogin
   */
  const verifyLogin = async () => {
    console.log('verify start');
    try {
      const resp = await authService.verify({});
      if (resp.userName) {
        await getMyUserInfo(resp.userName);
      }
    }
    catch (error) {
      setLoginUser(undefined);
      //handleError(error);
    }
    finally {
      console.log('verify end');
    }
  }


  /**
   * login: SignIn 
   */
  const login = async (userName: string, password: string) => {
    console.log('login');
    try {
      const res = await authService.login({ userName: userName, password: password });
      await getMyUserInfo(userName);
      console.log('login end');
      return res;
    }
    catch (error) {
      setLoginUser(undefined);
      console.log('login error');
      handleError(error);
      throw error;
    }
  }


  /**
   * login: MyUserInfo
   */
  const getMyUserInfo = async (userName: string) => {
    console.log('getMyUserInfo', userName);
    // if (loginUser || !username) {
    //   console.log('getMyUserInfo cancel', loginUser, username);
    //   return;
    // }
    try {
      const responseUser = await userService.getUser({ userName: userName });
      setLoginUser(prev => { console.log('setLoginUser', prev, responseUser.user); return responseUser.user! });
    }
    catch (error) {
      console.log('getMyUserInfo error');
      handleError(error);
      throw error;
    }
  }

  /**
   * login: MyUserInfo
   */
  const refreshUserInfo = async () => {
    console.log('refreshUserInfo');
    if (loginUser) {
      getMyUserInfo(loginUser.name);
    }
  }

  /**
   * logout: Pagetemplate 
   */
  const logout = async () => {
    console.log('logout');
    try {
      await authService.logout({});
    }
    catch (error) {
      console.log('logout error');
      handleError(error);
      throw error;
    }
    finally {
      console.log('logout end');
      setLoginUser(prev => { console.log('setLoginUser', `${prev} -> undefined`); return undefined });
    }
  }

  /**
   * updataPassword
   */
  const updataPassword = async (currentPassword: string, newPassword: string) => {
    console.log('updataPassword', loginUser?.name);
    setMask();
    try {
      try {
        return await userService.updateUserPassword({ userName: loginUser!.name, currentPassword, newPassword });
      }
      catch (error) {
        handleError(error);
        throw error;
      }
    }
    finally {
      console.log('updataPassword end');
      releaseMask();
    }
  }

  /**
   * clearLoginUser 
   */
  const clearLoginUser = async () => {
    setLoginUser(undefined);
  }

  /**
   * error handler
   */
  const handleError = (error: any) => {
    console.log(error);
    const msg = error?.message;
    msg && enqueueSnackbar(msg, { variant: 'error' });
  }

  return {
    loginUser,
    verifyLogin,
    login,
    logout,
    refreshUserInfo,
    updataPassword,
    clearLoginUser,
  };
}

/**
 * Provider
 */
export const LoginProvider: React.FC<React.PropsWithChildren<unknown>> = ({ children }) => {
  console.log('LoginProvider');
  const loginModule = useLoginModule();
  const [isVerified, setIsVerified] = useState(false);

  useEffect(() => {
    loginModule.verifyLogin()
      .then(() => setIsVerified(true));
  }, []); // eslint-disable-line

  return (
    <Context.Provider value={loginModule}>
      {isVerified ? children :
        <div>
          <CssBaseline />
          <Stack sx={{ m: 10 }} alignItems='center' spacing={2}>
            <Box sx={{ display: 'flex' }}>
              <CircularProgress />
            </Box>
            <Typography >verifying session...</Typography>
          </Stack>
        </div>}
    </Context.Provider >
  )
}

/**
 * export private member. (for test)
 */
// export const __local__ = {
//   useLoginModule,
// };
