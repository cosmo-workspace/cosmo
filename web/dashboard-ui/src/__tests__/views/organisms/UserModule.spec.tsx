import { Code, ConnectError } from '@bufbuild/connect';
import '@testing-library/jest-dom';
import { act, cleanup, renderHook } from '@testing-library/react';
import { useSnackbar } from "notistack";
import React from 'react';
import { afterEach, beforeEach, describe, expect, it, MockedFunction, vi } from "vitest";
import { useLogin } from '../../../components/LoginProvider';
import { useProgress } from '../../../components/ProgressProvider';
import { Template } from '../../../proto/gen/dashboard/v1alpha1/template_pb';
import { GetUserAddonTemplatesResponse } from '../../../proto/gen/dashboard/v1alpha1/template_service_pb';
import { User } from '../../../proto/gen/dashboard/v1alpha1/user_pb';
import { CreateUserResponse, DeleteUserResponse, GetUsersResponse, UpdateUserDisplayNameResponse, UpdateUserRoleResponse } from '../../../proto/gen/dashboard/v1alpha1/user_service_pb';
import { useTemplateService, useUserService } from "../../../services/DashboardServices";
import { UserContext, useTemplates, useUserModule } from '../../../views/organisms/UserModule';

//--------------------------------------------------
// mock definition
//--------------------------------------------------
vi.mock('notistack');
vi.mock('../../../components/LoginProvider');
vi.mock('../../../services/DashboardServices');
vi.mock('../../../components/ProgressProvider');
vi.mock('react-router-dom', () => ({
  useNavigate: () => vi.fn(),
}));

type MockedMemberFunction<T extends (...args: any) => any> = {
  [P in keyof ReturnType<T>]: MockedFunction<ReturnType<T>[P]>;
};

const useUserServiceMock = useUserService as MockedFunction<typeof useUserService>;
const userMock: MockedMemberFunction<typeof useUserService> = {
  getUser: vi.fn(),
  getUsers: vi.fn(),
  deleteUser: vi.fn(),
  createUser: vi.fn(),
  updateUserDisplayName: vi.fn(),
  updateUserPassword: vi.fn(),
  updateUserRole: vi.fn(),
}
const useLoginMock = useLogin as MockedFunction<typeof useLogin>;
const loginMock: MockedMemberFunction<typeof useLogin> = {
  loginUser: {} as any,
  verifyLogin: vi.fn(),
  login: vi.fn(),
  logout: vi.fn(),
  updataPassword: vi.fn(),
  refreshUserInfo: vi.fn(),
  clearLoginUser: vi.fn(),
};
const useTemplateServiceMock = useTemplateService as MockedFunction<typeof useTemplateService>;
const templateMock: MockedMemberFunction<typeof useTemplateService> = {
  getUserAddonTemplates: vi.fn(),
  getWorkspaceTemplates: vi.fn(),
}

const useProgressMock = useProgress as MockedFunction<typeof useProgress>;
const progressMock: MockedMemberFunction<typeof useProgress> = {
  setMask: vi.fn(),
  releaseMask: vi.fn(),
}
const useSnackbarMock = useSnackbar as MockedFunction<typeof useSnackbar>;
const snackbarMock: MockedMemberFunction<typeof useSnackbar> = {
  enqueueSnackbar: vi.fn(),
  closeSnackbar: vi.fn(),
}

//--------------------------------------------------
// mock data definition
//--------------------------------------------------
const user1 = new User({ name: 'user1', roles: ['cosmoAdmin'], displayName: 'user1 name' });
const user2 = new User({ name: 'user2', displayName: 'user2 name' });
const user3 = new User({ name: 'user3', displayName: 'user3 name' });

//-----------------------------------------------
// test
//-----------------------------------------------
describe('useUserModule', () => {

  beforeEach(async () => {
    useSnackbarMock.mockReturnValue(snackbarMock);
    useProgressMock.mockReturnValue(progressMock);
    useUserServiceMock.mockReturnValue(userMock);
    useLoginMock.mockReturnValue(loginMock);
  });

  afterEach(() => {
    expect(snackbarMock.enqueueSnackbar.mock.calls).toMatchSnapshot();
    vi.restoreAllMocks();
    cleanup();
  });

  async function renderUseUserModule() {
    return renderHook(() => useUserModule(), {
      wrapper: ({ children }) => (<UserContext.Provider>{children}</UserContext.Provider>),
    });
  }

  describe('useUserModule getUsers', () => {

    it('normal', async () => {
      const { result } = await renderUseUserModule();
      userMock.getUsers.mockResolvedValue(new GetUsersResponse({ items: [user2, user1, user3] }));
      await act(async () => { result.current.getUsers() });
      expect(result.current.users).toMatchSnapshot();
    });

    it('normal2', async () => {
      const { result } = await renderUseUserModule();
      userMock.getUsers.mockResolvedValue(new GetUsersResponse({ items: undefined as any }));
      await act(async () => { result.current.getUsers() });
      expect(result.current.users).toMatchSnapshot();
    });

    it('normal3', async () => {
      const { result } = await renderUseUserModule();
      userMock.getUsers.mockResolvedValue(new GetUsersResponse({ items: [] }));
      await act(async () => { result.current.getUsers() });
      expect(result.current.users).toMatchSnapshot();
    });

    it('error', async () => {
      const { result } = await renderUseUserModule();
      //userMock.getUsers.mockRejectedValue(new Error('[mock] getUsers error'));
      userMock.getUsers.mockRejectedValue(new ConnectError('[mock] getUsers error', Code.Unauthenticated));
      await expect(result.current.getUsers()).rejects.toMatchSnapshot();
    });

  });


  describe('useUserModule createUser', () => {
    beforeEach(async () => {
      userMock.getUsers.mockResolvedValue(new GetUsersResponse({ items: [user1, user2, user3] }));
    });

    it('nomal', async () => {
      const { result } = await renderUseUserModule();
      await act(async () => { result.current.getUsers() });
      userMock.createUser.mockResolvedValue(new CreateUserResponse({ message: "ok", user: user2 }));
      await act(async () => { result.current.createUser('user2', 'user2 name') });
      expect(result.current.users).toMatchSnapshot();
    });

    it('error', async () => {
      const { result } = await renderUseUserModule();
      await act(async () => { result.current.getUsers() });
      userMock.createUser.mockRejectedValue(new Error('[mock] createUser error'));
      await expect(result.current.createUser('user2', 'user2 name')).rejects.toMatchSnapshot();
    });
  });

  describe('useUserModule updateUserName', () => {

    it('nomal before getUsers', async () => {
      const { result } = await renderUseUserModule();
      const user2x = { ...user2, displayName: 'displayNameChange' }
      userMock.updateUserDisplayName.mockResolvedValue(new UpdateUserDisplayNameResponse({ message: "ok", user: user2x }));
      await act(async () => { result.current.updateName('user2', 'displayNameChange') });
      expect(result.current.users).toMatchSnapshot();
    });

    it('nomal after getUsers', async () => {
      const { result } = await renderUseUserModule();
      userMock.getUsers.mockResolvedValue(new GetUsersResponse({ message: "ok", items: [user1, user2, user3] }));
      await act(async () => { result.current.getUsers() });
      const user2x = { ...user2, displayName: 'displayNameChange' }
      userMock.updateUserDisplayName.mockResolvedValue(new UpdateUserDisplayNameResponse({ message: "ok", user: user2x }));
      await act(async () => { result.current.updateName('user2', 'displayNameChange') });
      expect(result.current.users).toMatchSnapshot();
    });

    it('nomal after getUsers. return value nothing', async () => {
      const { result } = await renderUseUserModule();
      userMock.getUsers.mockResolvedValue(new GetUsersResponse({ message: "ok", items: [user1, user2, user3] }));
      await act(async () => { result.current.getUsers() });
      userMock.updateUserDisplayName.mockResolvedValue(new UpdateUserDisplayNameResponse({ message: "ok", user: undefined as any }));
      await act(async () => { result.current.updateName('user2', 'displayNameChange') });
      expect(result.current.users).toMatchSnapshot();
    });

    it('error', async () => {
      const { result } = await renderUseUserModule();
      userMock.updateUserDisplayName.mockRejectedValue(new Error('[mock] putUserName error'));
      await expect(result.current.updateName('user2', 'displayNameChange')).rejects.toMatchSnapshot();
    });
  });

  describe('useUserModule updateRole', () => {

    it('nomal before getUsers', async () => {
      const { result } = await renderUseUserModule();
      userMock.updateUserRole.mockResolvedValue(new UpdateUserRoleResponse({ message: "ok", user: user2 }));
      await act(async () => { result.current.updateRole('user2', ['user2 name']) });
      expect(result.current.users).toMatchSnapshot();
    });

    it('nomal after getUsers', async () => {
      const { result } = await renderUseUserModule();
      userMock.getUsers.mockResolvedValue(new GetUsersResponse({ message: "ok", items: [user1, user2, user3] }));
      await act(async () => { result.current.getUsers() });

      const user2x = { ...user2, roles: ['Role2'] }
      userMock.updateUserRole.mockResolvedValue(new UpdateUserRoleResponse({ message: "ok", user: user2x }));

      await act(async () => { result.current.updateRole('user2', ['Role2']) });
      expect(result.current.users).toMatchSnapshot();
    });

    it('nomal after getUsers. return value nothing', async () => {
      const { result } = await renderUseUserModule();
      userMock.getUsers.mockResolvedValue(new GetUsersResponse({ message: "ok", items: [user1, user2, user3] }));
      await act(async () => { result.current.getUsers() });

      userMock.updateUserRole.mockResolvedValue(new UpdateUserRoleResponse({ message: "ok", user: undefined as any }));

      await act(async () => { result.current.updateRole('user2', ['Role2']) });
      expect(result.current.users).toMatchSnapshot();
    });

    it('error', async () => {
      const { result } = await renderUseUserModule();
      userMock.updateUserRole.mockRejectedValue(new Error('[mock] updateUserRole error'));
      await expect(result.current.updateRole('user2', ['user2 name'])).rejects.toMatchSnapshot();
    });
  });

  describe('useUserModule deleteUser', () => {
    beforeEach(async () => {
      userMock.getUsers.mockResolvedValue(new GetUsersResponse({ message: "ok", items: [user1, user2, user3] }));
    });

    it('nomal', async () => {
      const { result } = await renderUseUserModule();
      await act(async () => { result.current.getUsers() });
      userMock.deleteUser.mockResolvedValue(new DeleteUserResponse({ message: "ok", user: user2 }));
      await act(async () => { result.current.deleteUser('user2') });
      expect(result.current.users).toMatchSnapshot();
    });

    it('error', async () => {
      const { result } = await renderUseUserModule();
      await act(async () => { result.current.getUsers() });
      userMock.deleteUser.mockRejectedValue(new Error('[mock] deleteUser error'));
      await expect(result.current.deleteUser('user2')).rejects.toMatchSnapshot();
    });
  });

});

describe('useTemplates', () => {

  beforeEach(async () => {
    useSnackbarMock.mockReturnValue(snackbarMock);
    useProgressMock.mockReturnValue(progressMock);
    useTemplateServiceMock.mockReturnValue(templateMock);
    useLoginMock.mockReturnValue(loginMock);
  });

  afterEach(() => {
    vi.restoreAllMocks();
    cleanup();
  });

  async function renderUseTemplates() {
    const utils = renderHook(() => useTemplates(), {
      wrapper: ({ children }) => (<UserContext.Provider>{children}</UserContext.Provider>),
    });
    return utils;
  }

  describe('useTemplates getUserAddonTemplates', () => {

    const tmpl1 = new Template({ name: 'tmpl1' });
    const tmpl2 = new Template({
      name: 'tmpl2',
      description: "hoge",
      requiredVars: [{ varName: 'var1', defaultValue: 'var1Value' }, { varName: 'var2' }],
      isDefaultUserAddon: true,
    });
    const tmpl3 = new Template({ name: 'tmpl3' });

    it('normal', async () => {
      const { result } = await renderUseTemplates();
      templateMock.getUserAddonTemplates.mockResolvedValue(new GetUserAddonTemplatesResponse({
        message: "ok", items: [tmpl1, tmpl3, tmpl2]
      }));
      await act(async () => { result.current.getUserAddonTemplates() });
      expect(result.current.templates).toMatchSnapshot();
    });

    it('normal template empty', async () => {
      const { result } = await renderUseTemplates();
      templateMock.getUserAddonTemplates.mockResolvedValue(new GetUserAddonTemplatesResponse({ items: [] }));
      await act(async () => { result.current.getUserAddonTemplates() });
      expect(result.current.templates).toMatchSnapshot();
    });

    it('error', async () => {
      const { result } = await renderUseTemplates();
      templateMock.getUserAddonTemplates.mockRejectedValue(new Error('[mock] getUser error'));
      await expect(result.current.getUserAddonTemplates()).rejects.toMatchSnapshot();
      expect(result.current.templates).toMatchSnapshot();
    });

  });

});
