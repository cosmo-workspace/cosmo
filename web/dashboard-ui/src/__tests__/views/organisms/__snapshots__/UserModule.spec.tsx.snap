// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useTemplates > useTemplates getUserAddonTemplates > error 1`] = `[Error: [mock] getUser error]`;

exports[`useTemplates > useTemplates getUserAddonTemplates > error 2`] = `[]`;

exports[`useTemplates > useTemplates getUserAddonTemplates > normal 1`] = `
[
  {
    "description": "",
    "isClusterScope": false,
    "name": "tmpl1",
    "requiredVars": [],
  },
  {
    "description": "hoge",
    "isClusterScope": false,
    "isDefaultUserAddon": true,
    "name": "tmpl2",
    "requiredVars": [
      {
        "defaultValue": "var1Value",
        "varName": "var1",
      },
      {
        "defaultValue": "",
        "varName": "var2",
      },
    ],
  },
  {
    "description": "",
    "isClusterScope": false,
    "name": "tmpl3",
    "requiredVars": [],
  },
]
`;

exports[`useTemplates > useTemplates getUserAddonTemplates > normal template empty 1`] = `[]`;

exports[`useUserModule > useUserModule createUser > error 1`] = `[Error: [mock] createUser error]`;

exports[`useUserModule > useUserModule createUser > error 2`] = `
[
  [
    "[mock] createUser error",
    {
      "variant": "error",
    },
  ],
]
`;

exports[`useUserModule > useUserModule createUser > nomal 1`] = `
[
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user1 name",
    "name": "user1",
    "roles": [
      "cosmoAdmin",
    ],
    "status": "",
  },
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user2 name",
    "name": "user2",
    "roles": [],
    "status": "",
  },
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user3 name",
    "name": "user3",
    "roles": [],
    "status": "",
  },
]
`;

exports[`useUserModule > useUserModule createUser > nomal 2`] = `
[
  [
    "ok",
    {
      "variant": "success",
    },
  ],
]
`;

exports[`useUserModule > useUserModule deleteUser > error 1`] = `[Error: [mock] deleteUser error]`;

exports[`useUserModule > useUserModule deleteUser > error 2`] = `
[
  [
    "[mock] deleteUser error",
    {
      "variant": "error",
    },
  ],
]
`;

exports[`useUserModule > useUserModule deleteUser > nomal 1`] = `
[
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user1 name",
    "name": "user1",
    "roles": [
      "cosmoAdmin",
    ],
    "status": "",
  },
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user3 name",
    "name": "user3",
    "roles": [],
    "status": "",
  },
]
`;

exports[`useUserModule > useUserModule deleteUser > nomal 2`] = `
[
  [
    "ok",
    {
      "variant": "success",
    },
  ],
]
`;

exports[`useUserModule > useUserModule getUsers > error 1`] = `[ConnectError: [unauthenticated] [mock] getUsers error]`;

exports[`useUserModule > useUserModule getUsers > error 2`] = `
[
  [
    "[unauthenticated] [mock] getUsers error",
    {
      "variant": "error",
    },
  ],
]
`;

exports[`useUserModule > useUserModule getUsers > normal 1`] = `
[
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user1 name",
    "name": "user1",
    "roles": [
      "cosmoAdmin",
    ],
    "status": "",
  },
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user2 name",
    "name": "user2",
    "roles": [],
    "status": "",
  },
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user3 name",
    "name": "user3",
    "roles": [],
    "status": "",
  },
]
`;

exports[`useUserModule > useUserModule getUsers > normal 2`] = `[]`;

exports[`useUserModule > useUserModule getUsers > normal2 1`] = `[]`;

exports[`useUserModule > useUserModule getUsers > normal2 2`] = `[]`;

exports[`useUserModule > useUserModule getUsers > normal3 1`] = `[]`;

exports[`useUserModule > useUserModule getUsers > normal3 2`] = `[]`;

exports[`useUserModule > useUserModule updateRole > error 1`] = `[Error: [mock] updateUserRole error]`;

exports[`useUserModule > useUserModule updateRole > error 2`] = `
[
  [
    "[mock] updateUserRole error",
    {
      "variant": "error",
    },
  ],
]
`;

exports[`useUserModule > useUserModule updateRole > nomal after getUsers 1`] = `
[
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user1 name",
    "name": "user1",
    "roles": [
      "cosmoAdmin",
    ],
    "status": "",
  },
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user2 name",
    "name": "user2",
    "roles": [
      "Role2",
    ],
    "status": "",
  },
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user3 name",
    "name": "user3",
    "roles": [],
    "status": "",
  },
]
`;

exports[`useUserModule > useUserModule updateRole > nomal after getUsers 2`] = `
[
  [
    "ok",
    {
      "variant": "success",
    },
  ],
]
`;

exports[`useUserModule > useUserModule updateRole > nomal after getUsers. return value nothing 1`] = `
[
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user1 name",
    "name": "user1",
    "roles": [
      "cosmoAdmin",
    ],
    "status": "",
  },
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user2 name",
    "name": "user2",
    "roles": [],
    "status": "",
  },
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user3 name",
    "name": "user3",
    "roles": [],
    "status": "",
  },
]
`;

exports[`useUserModule > useUserModule updateRole > nomal after getUsers. return value nothing 2`] = `
[
  [
    "ok",
    {
      "variant": "success",
    },
  ],
]
`;

exports[`useUserModule > useUserModule updateRole > nomal before getUsers 1`] = `[]`;

exports[`useUserModule > useUserModule updateRole > nomal before getUsers 2`] = `
[
  [
    "ok",
    {
      "variant": "success",
    },
  ],
]
`;

exports[`useUserModule > useUserModule updateUserName > error 1`] = `[Error: [mock] putUserName error]`;

exports[`useUserModule > useUserModule updateUserName > error 2`] = `
[
  [
    "[mock] putUserName error",
    {
      "variant": "error",
    },
  ],
]
`;

exports[`useUserModule > useUserModule updateUserName > nomal after getUsers 1`] = `
[
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user1 name",
    "name": "user1",
    "roles": [
      "cosmoAdmin",
    ],
    "status": "",
  },
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "displayNameChange",
    "name": "user2",
    "roles": [],
    "status": "",
  },
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user3 name",
    "name": "user3",
    "roles": [],
    "status": "",
  },
]
`;

exports[`useUserModule > useUserModule updateUserName > nomal after getUsers 2`] = `
[
  [
    "ok",
    {
      "variant": "success",
    },
  ],
]
`;

exports[`useUserModule > useUserModule updateUserName > nomal after getUsers. return value nothing 1`] = `
[
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user1 name",
    "name": "user1",
    "roles": [
      "cosmoAdmin",
    ],
    "status": "",
  },
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user2 name",
    "name": "user2",
    "roles": [],
    "status": "",
  },
  {
    "addons": [],
    "authType": "",
    "defaultPassword": "",
    "displayName": "user3 name",
    "name": "user3",
    "roles": [],
    "status": "",
  },
]
`;

exports[`useUserModule > useUserModule updateUserName > nomal after getUsers. return value nothing 2`] = `
[
  [
    "ok",
    {
      "variant": "success",
    },
  ],
]
`;

exports[`useUserModule > useUserModule updateUserName > nomal before getUsers 1`] = `[]`;

exports[`useUserModule > useUserModule updateUserName > nomal before getUsers 2`] = `
[
  [
    "ok",
    {
      "variant": "success",
    },
  ],
]
`;
